File: Java/Subsets.java
class Solution {

    private void helper(int[] nums, int start,
                       List<Integer> elem, List<List<Integer>> result) {
        result.add(new ArrayList<>(elem));
        for(int i = start; i < nums.length; i++) {
            elem.add(nums[i]);
            helper(nums, i+1, elem, result);
            elem.remove(elem.size() - 1);
        }
  }

    public List<List<Integer>> subsets(int[] nums) {
        if (nums == null || nums.length == 0) return new ArrayList<>();

        List<List<Integer>> res = new ArrayList<>();
        List<Integer> elem = new ArrayList<>();
        helper(nums, 0, elem, res);

        return res;
    }
}

File: Go/subsets.go
func helper(nums []int, start int, elem []int, res *[][]int) {
    *res = append(*res, append([]int{}, elem...))
    for i := start; i < len(nums); i += 1 {
        elem = append(elem, nums[i])
        helper(nums, i + 1, elem, res)
        elem = elem[:len(elem) - 1]
    }
}

func subsets(nums []int) [][]int {
    var res [][]int
    var elem []int
    helper(nums, 0, elem, &res)
    return res
}